### PR TITLE
fix stop_str; response_length; add batch size in config

### PR DESCRIPTION
**Pull request by xzzWZY**

### CODE REVIEW

**Review by jaywonchung — CHANGES_REQUESTED** [review]
Okay. I can't understand the code within a reasonable amount of time. Very few people will be able to.

We're not writing this script to just get it working now and forget about it. If that was the purpose, we would not have open sourced the codebase. We're rather aiming to provide people with a reasonable script to reproduce the benchmark and allow them to potentially build their own thing on top of this, and this requires good code quality on our side. Another thing is that code of this level of complexity will surely have bugs that are difficult to catch. 

Due to the Chameleon Cloud cluster outage, we have time until the 31st. This is precisely the time to refactor this script. The goal is to recover code quality to what it used to be like [here](https://github.com/ml-energy/leaderboard/blob/ef7e22b16a3420a428a4caf274ab7a3f3bdda96e/scripts/benchmark.py).

Some comments to get you started:
1. The part of `main` that uses `generate_stream` is needlessly complicated, with unclear temporary variables like `stop_token`, `stop_str`, `batch_token_len`, etc. Initially, before batching, using `generate_stream` from FastChat made sense because there was just one *stream* of tokens for the single input prompt generated from the model. Now, the benchmarking script is running multiple prompts at the same time, and the *stream* abstraction no longer makes sense. That's why the code that uses `generate_stream` has become overcomplicated -- because it has to twist and bend stuff to make the wrong abstraction work.
2. Rewrite `generate_stream` to `def run_inference(model, tokenizer, gen_params, device, context_len) -> list[Output]` where `Output` is a Python [`dataclass`](https://docs.python.org/3/library/dataclasses.html) that holds values `response_length`, `input`, and `output`. In `main`, `monitor.begin_window` and `monitor.end_window` will just wrap around `run_inference`. Then, it'll calculate `latency`,` energy`, and `throughput` using `Measurements` and save those with `model_name_cleaned` and the values inside `Output`  as one entry in `output_json`. It will also print out prompts and model responses with `console.print`.
4. Remove all unnecessary logic that used to exist in `generate_stream`, for instance, checking if `device == "mps"`. `generate_stream` (or `run_inference`) will no longer `yield` anything. Rather, it will encapsulate the dirty parts of running the model, and then simply `return` one `list[Output]` at the end.
5. Now that I look into the code, you're using `np.vectorize` to *"vectorize"*, but that is nothing but syntactic sugar. The [documentation](https://numpy.org/doc/stable/reference/generated/numpy.vectorize.html) also says that `The vectorize function is provided primarily for convenience, not for performance. The implementation is essentially a for loop.`. Just forget about vectorizing code and revert back to a naive for loop over all the responses to determine (1) which token index marks the end of each response and (2) whether every response is done -- performance will be identical to what you have here, but with a large for loop over all responses, at least the code will be more readable.

This is what I think makes sense, before writing it myself. As you refactor, some other design might make more sense to you. Trust your instinct and we should definitely talk about it.

**Comment by xzzWZY** [comment]
I totally agree with your design.

A tiny question, we still need to wait all requests in a batch to finish generating tokens, and then return the list[Output], which means the requests stopped earlier will continue to generate some random things and we only return where they should stop in list[Output], right?

**Comment by jaywonchung** [comment]
Yep, and if I understood correctly, we were already doing that, weren't we? And yeah, even some responses stop earlier and generate random tokens, the corresponding `Output.output` field will only contain *valid words* and `Output.response_length` will only count the number of *valid tokens*.

**Comment by xzzWZY** [comment]
```
        if temperature < 1e-5 or top_p < 1e-8:  # greedy
            _, indices = torch.topk(last_token_logits, 1)
            tokens = [[int(token) for token in query] for query in indices.tolist()]
        else:
            probs = torch.softmax(last_token_logits, dim=-1)
            indices = torch.multinomial(probs, num_samples=1)
            tokens = [[int(token) for token in query] for query in indices.tolist()]

```
here previously we use num_samples=2, inherit from fastchat. fastchat use num_samples=2 because they have an option to change the next token to second one from samples and take the whole output_ids as input to decoder layer, in order to always generate maximum length.

We decide not to consider this situation, so that it's enough to use num_samples = 1, but previous benchmark evaluation uses fastchat.generate_stream, which use num_samples = 2, is it fine to just change to num_samples = 1 here?

**Comment by jaywonchung** [comment]
Why should we change to `num_samples=1` when `num_samples=2` works fine?

**Comment by xzzWZY** [comment]
Both works, but I think if we only need one token for each iteration, we don't need to set num_samples = 2

**Comment by jaywonchung** [comment]
If you're sure both works it's fine to change. But only change if you're 100% sure you're not introducing a bug or a significant code behavior change. The performance gain will be marginal anyway and that's why I thought it's not worth the risk.

**Comment by xzzWZY** [comment]
I'm sure it'll not introduce bug, but I'm almost sure it will introduce a performance change by taking only one sample instead of two. To keep batch inference consistent with previous results using generate_stream, I guess it's better to change back to num_samples = 2

**Comment by xzzWZY** [comment]
when I add llama 2 A100/V100 batch 1 to benchmark, I find there's a significant increase in throughput compared to llama, e.g.

**for A100**

previous(using original fastchat in benchmark)
model | throughput | response_length | latency | energy
-- | -- | -- | -- | --
metaai/llama-13B | 23.201365367977594 | 76.47078576225655 | 2.9742001407414107 | 631.5696930826057
metaai/llama-7B | 27.470901624921254 | 60.65144392209537 | 1.9862404549640509 | 335.25658764271293

using modified run_reference
model | throughput | response_length | latency | energy
-- | -- | -- | -- | --
metaai/Llama-2-13b-chat-hf | 30.00588500993857 | 379.4936198791135 | 12.98423684546037 | 3050.1256225654797
metaai/Llama-2-7b-chat-hf | 44.46873568590649 | 420.0661517797179 | 9.4420965670117 | 1828.7749392209557

**for V100**
previous(using original fastchat in benchmark)
model | throughput | response_length | latency | energy
-- | -- | -- | -- | --
metaai/llama-13B | 17.273409248575422 | 76.56447280053727 | 3.7837350834271555 | 772.3963885157816
metaai/llama-7B | 26.078374550887073 | 65.01040967092008 | 2.106037566082945 | 335.8834066487579

using modified run_reference
model | throughput | response_length | latency | energy
-- | -- | -- | -- | --
metaai/Llama-2-13b-chat-hf | 20.34702087033603 | 374.93485560779044 | 18.96952100525018 | 3569.768710208201
metaai/Llama-2-7b-chat-hf | 36.61609650432256 | 419.32471457353927 | 11.42535028784149 | 2110.3780402954935

while for A40, the increase is not so large, e.g.
previous( using original fastchat in benchmark)
model | throughput | response_length | latency | energy
-- | -- | -- | -- | --
metaai/llama-13B | 15.699146010424393 | 80.32236400268637 | 4.757332595030835 | 1293.689832437891
metaai/llama-7B | 25.768609507174105 | 64.59032907991941 | 2.284814629996714 | 525.7081235728675

using modeified generate_stream to support batched inference, but still have same structure as fastchat
model | throughput | response_length | latency | energy
-- | -- | -- | -- | --
metaai/Llama-2-13b-chat-hf | 16.95804416983929 | 384.7333781061115 | 22.55271715111622 | 4337.670243116255
metaai/Llama-2-7b-chat-hf | 31.922994116700572 | 428.19341840161184 | 13.367807321468502 | 2556.7166067830576


One possible reason is that when I modify the benchmark.py by re-structuring the code from generate_stream to run_reference (and using dataclass to deal with edge case), the generate process becomes more effcient, and it automatically increase throughput.

To verfy if re-structuring the code really increase the throughput, I run llama-7B on ampere01 again, and find throughput indeed increase

model | throughput | response_length | latency | energy
-- | -- | -- | -- | --
metaai/llama-7B | 25.768609507174105 | 64.59032907991941 | 2.284814629996714 | 525.7081235728675
metaai/llama-7B | 28.747727704665646 | 65.16823371390194 | 2.3498663063789045 | 568.1551541302771

where the first line is generated by previous generate_stream, second line is generated by run_reference

but actually I find if we get the total throughput by response_length/latency, where response_length and latency are both average among all queries, (currently we calculate throughput by averaging throughput of all queries) we can find these two throughput is quite close to each other:
previous generate_stream | run_inference
-- | --
28.2693958 | 27.732741

the reason is currently we calculate the throughput by averaging the throughput of each single inference, but the new method is to use tokens generate in all all inference to divide total latency in all inference. I think the former one makes more sense in user level, since this indicates the average generation speed (#token / second) from users' prespective, and the latter one makes more sense in system level, since it indicates the total throughput when running a large amount of queries, and it's more stable when response length varies for each query.

Thus, I'm still not sure if re-structuring the code inflence efficiency, but at least the small experiment on ampere node doesn't show it. The difference of throughput in small experiment mostly because of randomness of response_length for each query, I think, since the total throughput = total tokens / total latnecy are close; 

and another question is if we need to change the method to calculate throughput from averaging throughput for each inference to total tokens / total latency. For models which are verbal like llama 2, these two numbers are close, while for models which are not verbal like llama, these two numbers are more different.

**Comment by jaywonchung** [comment]
Thank you for the comprehensive benchmark and comparison! This is really helpful!
I didn't know that generators had such a large performance hit to text generation, but it also makes sense.

I think this should mark a new release of the leaderboard. That is, we should run batch size = 1 and that result (with improved throughput and different energy consumption) will replace the data we previously had. Then, we'll release that as leaderboard V2, together with other batch sizes.

A comment regarding the llama-2 data you added -- What is the batch size?
In leaderboard V2 with the improved benchmark script, we should first find the maximum batch size for the model, which will be different for each GPU (because they have different amounts of memory). Then, run three batch sizes: 1, max, and (max + 1) // 2. There will be a new `batch size` column in the leaderboard table.

After you find the maximum batch size, please try to make all of these just one Pegasus yaml file, so that you don't have to spend too much time running these. Also, let me know if you need any more time on resources -- I'll try to figure out.

**Comment by xzzWZY** [comment]
> A comment regarding the llama-2 data you added -- What is the batch size?

currently I just add batch size = 1, only batch size = 1 are done. I just let llama 2 batch size = 1 

(1) to make myself clear, you mean, do you mean that I don't need to run fix batch size = 2,4,8 on A100 and V100 now (just stop it), but doing this 

> In leaderboard V2 with the improved benchmark script, we should first find the maximum batch size for the model, which will be different for each GPU (because they have different amounts of memory). Then, run three batch sizes: 1, max, and (max + 1) // 2. There will be a new batch size column in the leaderboard table.

(2) and I currently don't come up with a way to quickly decide the maximum batch size for each model. Should we search on web to get the model weight and size of kv-cache per token for each model? or should we get it by running the model with batch size 1 and let it generate maximum length and during this process, we measure the memory consumption and then decide maximum batch? and how can we convert it into one yaml file?

(3) for the two method of calculating throughput mentioned above, should we still use throughput = mean(throughput_{i}) or throughput = (mean(response_length))/(mean(lantency)) ?

**Comment by jaywonchung** [comment]
(1,2) Actually, what about we just run a grid of batch sizes (e.g., 1, 2, 4, 8, 16, 32, 64) with Pegasus, and some batch sizes will die of OOM (and `benchmark.json` will be incomplete). Then, we post whatever batch size that survived. WDYT?
(3) I think the former is correct. When people look at the `throughput` number, they would expect the number of tokens per seconds they would get if they run their prompt on it.

**Comment by xzzWZY** [comment]
> (1,2) Actually, what about we just run a grid of batch sizes (e.g., 1, 2, 4, 8, 16, 32, 64) with Pegasus, and some batch sizes will die of OOM (and benchmark.json will be incomplete). Then, we post whatever batch size that survived. WDYT?

I agree with doing so. Based on my observation, some 13B models dies incomplete on batch size = 2 for V100, since I sorted input length from short to long, they dies at about 2500+/2978; most 13B models survive for batch size = 4 on A40 and dies at batch size = 8; for A100, all survive for batch size = 8

Since we need to re-run batch size = 1 on three kinds of GPUs, I think I need more time on V100, and ampere node for re-run A40 batch 1, 2

**Comment by jaywonchung** [comment]
Ah, then we should the sort dataset in reverse order (longer prompt first) so that experiments will fail earlier, right?

**Comment by xzzWZY** [comment]
agree

File: scripts/benchmark.py [hunk 1]
```diff
@@ -141,91 +147,73 @@ def generate_stream(
...
+        output = tokenizer.batch_decode(
+            output_ids,
+            skip_special_tokens=True,
+            spaces_between_special_tokens=False,
+            clean_up_tokenization_spaces=True,
+        )
+        output_np = np.array(output)
+
+        # deal with stop_str
+        if stop_str:
```
Comment (jaywonchung) [inline scripts/benchmark.py:185]: Where does the code check for partially generated stop strings?

Comment (xzzWZY) [reply scripts/benchmark.py:185]: in benchmark we don't need to check the partially stop string, since in fastchat, it check the partially stop string because it needs to print every two tokens generated, while in benchmark we only print the final result, thus if there are partially stop string but not become the complete stop string, the refind will not find this as stop string

Comment (jaywonchung) [reply scripts/benchmark.py:185]: Ah that makes sense. Nice catch. Thanks for the explanation!

**Comment by jaywonchung** [comment]
@xzzWZY What is the status of this?

**Comment by xzzWZY** [comment]
Currently tasks on A40 are all finished, V100 finishes most of tasks, and A100 finishes about half

Additionally, I just found when running batch inference on llama 2-7b chat, it sometimes generates unexpected row of nan in logits = out.logits, which will further lead to a break of program in 
```
probs = torch.softmax(last_token_logits, dim=-1)
indices = torch.multinomial(probs, num_samples=2)
```
thus currently all the llama 2 7b chat only has result for batch = 1

this only happens for llama 2 7b, not for llama 2 13b and any other models. It's probably a problem for only llama 2 7b I guess, and I'm not sure if we need to modify code to avoid this happen, so currently I just remain the code unchange and let llama 2 7b died at some early place.

**Comment by jaywonchung** [comment]
The NaN issue is pretty strange, but I don't think it'll happen for every prompt/batch. When it happens, I think the benchmark script should just skip the batch and move on to the next one.

**Comment by jaywonchung** [comment]
@xzzWZY Can we just merge the code (`benchmark.py`) for now and add results later?
